internal/gengapic/genrest.go

```diff
@@ -56,62 +80,515 @@ func (g *generator) genRESTMethods(serv *descriptor.ServiceDescriptorProto, serv
-	// If we leave it commented out, we can change it, and we don't have to
-	// guarantee correctness or stability.
-	//
-	// p("// New%sRESTClient creates a new %s rest client.", servName, clientName)
-	// g.serviceDoc(serv)
-	// p("func New%[1]sRESTClient(ctx context.Context, opts ...option.ClientOption) (*%[1]sClient, error) {", servName)
-	// p("    c := &%s{", servName)
-	// p("    }")
-	// p("")
-	// p("    return &%sClient{internal%[1]sClient: c, CallOptions: default%[1]sCallOptions()}, nil", servName)
-	// p("}")
+func (g *generator) queryParams(m *descriptor.MethodDescriptorProto) map[string]*descriptor.FieldDescriptorProto {
+	queryParams := map[string]*descriptor.FieldDescriptorProto{}
+	info, err := getHTTPInfo(m)
+	if info == nil || err != nil {
+		return queryParams
+	}
+	pathParams := g.pathParams(m)
+	// Minor hack: we want to make sure that the body parameter, if one exists,
+	// is NOT a query parameter.
+	pathParams[info.body] = nil
+	msg := g.descInfo.Type[m.GetInputType()].(*descriptor.DescriptorProto)
+	if info.body != "" && info.body != "*" {
+		field := g.lookupField(m.GetInputType(), info.body)
+		msg = g.descInfo.Type[field.GetTypeName()].(*descriptor.DescriptorProto)
+	}
+
+	// If, and only if, a field is a path parameter, it's not a query parameter.
+	for _, f := range msg.Field {
+		if _, ok := pathParams[f.GetName()]; !ok {
+			queryParams[f.GetName()] = f
+		}
+	}
 
+	return queryParams
+}
+
+func (g *generator) generateQueryString(m *descriptor.MethodDescriptorProto) {
+	p := g.printf
+	queryParams := g.queryParams(m)
+	if queryParams == nil {
+		return
+	}
+
+	// We want to iterate over fields in a deterministic order
+	// to prevent spurious deltas when regenerating gapics.
+	fields := make([]string, 0, len(queryParams))
+	for p, _ := range queryParams {
+		fields = append(fields, p)
+	}
+	sort.Strings(fields)
+	info, _ := getHTTPInfo(m)
+
+	bodyStr := "req"
+	if info.body != "" && info.body != "*" {
+		bodyStr = "body"
+		p("body := req.Get%s()", snakeToCamel(info.body))
+	}
+	p("queryParamStrs := []string{}")
+	// TODO(dovs): make this work for nested fields
+	for _, f := range fields {
+		field := queryParams[f]
+		f = snakeToCamel(f)
+		// Note: something very strange has happened
+		if field.GetLabel() == descriptor.FieldDescriptorProto_LABEL_REPEATED {
+			// It's a slice, so check for nil
+			p("if %s.Get%s() != nil {", bodyStr, f)
+		} else {
+			// Default values are type specific
+			switch field.GetType() {
+			case descriptor.FieldDescriptorProto_TYPE_MESSAGE:
+				p("if %s.Get%s() != nil {", bodyStr, f)
+			case descriptor.FieldDescriptorProto_TYPE_BYTES:
+				p("if %s.Get%s() != nil {", bodyStr, f)
+			case descriptor.FieldDescriptorProto_TYPE_STRING:
+				p(`if %s.Get%s() != "" {`, bodyStr, f)
+			case descriptor.FieldDescriptorProto_TYPE_BOOL:
+				p(`if %s.Get%s() {`, bodyStr, f)
+			default: // Handles all numeric types
+				p(`if %s.Get%s() != 0 {`, bodyStr, f)
+			}
+		}
```

Comment:
I think I like multiple cases in case best. While definitely more idiomatic in go, I'm generally hesitant to use `break`, `continue`, or `fallthrough`. Trying to do something clever to make the switch handle both field type and label backfired, so I'll keep them separate.